include/seqan3/utility/views/join_with.hpp

```diff
@@ -6,26 +6,525 @@
+                                   && std::derived_from<std::forward_iterator_tag,
+                                                        typename std::iterator_traits<PatternIter>::iterator_category>,
+                               std::forward_iterator_tag,
+                               std::input_iterator_tag>>>;
+};
+
+template <bool Const, typename V, typename Pattern>
+struct iterator_category_t<Const, V, Pattern, false>
+{
+    using Parent = helper<Const, V, Pattern>::Parent;
+    using Base = helper<Const, V, Pattern>::Base;
+    using InnerBase = helper<Const, V, Pattern>::InnerBase;
+    using PatternBase = helper<Const, V, Pattern>::PatternBase;
+
+    using OuterIter = helper<Const, V, Pattern>::OuterIter;
+    using InnerIter = helper<Const, V, Pattern>::InnerIter;
+    using PatternIter = helper<Const, V, Pattern>::PatternIter;
+};
+
+} // namespace seqan3::detail::join_with
+
+namespace seqan3::detail
+{
+
+template <std::ranges::input_range V, std::ranges::forward_range Pattern>
+    requires std::ranges::view<V> && std::ranges::input_range<std::ranges::range_reference_t<V>>
+          && std::ranges::view<Pattern>
+          && join_with::compatible_joinable_ranges<std::ranges::range_reference_t<V>, Pattern>
+template <bool Const>
+class join_with_view<V, Pattern>::iterator :
+    public join_with::iterator_category_t<Const, V, Pattern, join_with::helper<Const, V, Pattern>::ref_is_glvalue>
+{
+    using helper_t = join_with::helper<Const, V, Pattern>;
+    using Parent = helper_t::Parent;
+    using Base = helper_t::Base;
+    using InnerBase = helper_t::InnerBase;
+    using PatternBase = helper_t::PatternBase;
+
+    using OuterIter = helper_t::OuterIter;
+    using InnerIter = helper_t::InnerIter;
+    using PatternIter = helper_t::PatternIter;
+
+    static constexpr bool ref_is_glvalue = helper_t::ref_is_glvalue;
+
+    friend class join_with_view<V, Pattern>;
+
+    Parent * parent_ = nullptr;
+
+public:
+    OuterIter outer_it_ = OuterIter();
+
+private:
+    std::variant<PatternIter, InnerIter> inner_it_;
+
+    constexpr iterator(Parent & parent, std::ranges::iterator_t<Base> outer) :
+        parent_(std::addressof(parent)),
+        outer_it_(std::move(outer))
+    {
+        if (outer_it_ != std::ranges::end(parent_->base_))
+        {
+            auto && inner = update_inner(outer_it_);
+            inner_it_.template emplace<1>(std::ranges::begin(inner));
+            satisfy();
+        }
+    }
+
+    constexpr auto && update_inner(OuterIter const & x)
+    {
+        if constexpr (ref_is_glvalue)
+            return *x;
+        else
+            return parent_->inner_.emplace_deref(x);
+    }
+
+    constexpr auto && get_inner(OuterIter const & x)
+    {
+        if constexpr (ref_is_glvalue)
+            return *x;
+        else
+            return *parent_->inner_;
+    }
+
+    constexpr void satisfy()
```

Comment:
can you add a short brief here? it's hard to review by just looking at the code.